File: dda.py
import pandas as pd

def dda(x1, x2, y1, y2):
    length = max(abs(x2 - x1), abs(y2 - y1))

    x_inc = (x2 - x1) / float(length)
    y_inc = (y2 - y1) / float(length)
    x, y = (x1, y1)

    data = []  # Create a list to store the data
    x_values = [] # Create a list to store the x values

    for i in range(1, length, 1):
        x_round = round(x)
        x_truncate = int(x)
        y_value = round(y)

        data.append((x_round, x_truncate, y_value, y1))
        x_values.append(x) # Append the x value to the list
        x += x_inc
        y += y_inc

    return data, x_values

import pandas as pd

data, x_values = dda(11, 4, 2, 25)

X_round, X_truncate, Y_values, Y1 = zip(*data)

df = pd.DataFrame({'x_values': x_values, 'X_round': X_round, 'X_truncate': X_truncate, 'Y_values': Y_values})

# df = df.style.format({'x': '{:.4f}', 'y': '{:.2f}', 'x_values': '{:.4f}'}).hide(axis='index')
print(df)

File: DDA/dda.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct
{
    /* data */
    int x;
    int x_round;
    int x_truncate;
    int y;
} Point;

Point* dda(int x1, int x2, int y1, int y2, int *length){
    *length = fmax(abs(x2 - x1), abs(y2 - y1));

    double x_inc = (double)(x2 - x1) / *length;
    double y_inc = (double)(y2 - y1) / *length;

    Point* points = (Point*)malloc(*length * sizeof(Point));

    double x = x1;
    double y = y1;

    for (int i =1; i < *length; ++i){
        int x_round = round(x);
        int x_truncate = (int)x;
        int y_value = round(y);


        points[i - 1].x = x_round;
        points[i - 1].x_round = x_round;
        points[i - 1].x_truncate = x_truncate;
        points[i - 1].y = y_value;

        x += x_inc;
        y += y_inc;

    }

    return points;
}


int main(){
    int x1 = 11, x2 = 4, y1 = 2, y2 = 25;

    int length;

    Point* result = dda(x1, x2, y1, y2, &length);

    // Display the table

    printf("x  | x_round | x_truncate | y \n");

    for (int i = 0; i < length - 1; ++i){
        printf("%2d | %7d | %10d | %d\n", result[i].x, result[i].x_round, result[i].x_truncate, result[i].y);
    }
    free(result);
    return 0;
}
